### PR TITLE
Feature/zimage inpaint pipeline

### PR DESCRIPTION
**Pull request by CalamitousFelicitousness**
# What does this PR do?

  This PR adds an inpainting pipeline for Z-Image. The summary of changes are below:
  - Implemented the ZImageInpaintPipeline class for mask-based inpainting
  - Updated the pipeline structure to include ZImageInpaintPipeline alongside ZImagePipeline and ZImageImg2ImgPipeline
  - Mapped the new ZImageInpaintPipeline in AUTO_INPAINT_PIPELINES_MAPPING
  - Added unit tests for ZImageInpaintPipeline with torch.empty fix for test stability
  - Updated dummy objects to include ZImageInpaintPipeline
  - Added documentation with usage example

Closes issue #12752

  Tested using a simple script:

<details>

<summary>Testing script</summary>

```python
  #!/usr/bin/env python
  """Test script for ZImage inpaint support."""

  import sys
  sys.path.insert(0, '/home/ohiom/diffusers/src')

  import torch
  import numpy as np
  from PIL import Image
  from diffusers import ZImageInpaintPipeline

  # Paths
  MODEL_PATH =
  "database/models/huggingface/models--Tongyi-MAI--Z-Image-Turbo/snapshots/78771b7e11b922c868dd766476bda1f4fc6bfc96"
  INPUT_IMAGE_PATH = "death_remix_1024.png"

  print("Loading ZImageInpaintPipeline...")
  pipe = ZImageInpaintPipeline.from_pretrained(
      MODEL_PATH,
      torch_dtype=torch.bfloat16,
      local_files_only=True,
  )
  pipe.to("cuda")
  print("Pipeline loaded.")

  # Load input image
  print(f"\nLoading input image from {INPUT_IMAGE_PATH}...")
  input_image = Image.open(INPUT_IMAGE_PATH).convert("RGB")
  print(f"Input image size: {input_image.size}")

  # Create a mask (white = inpaint, black = preserve)
  width, height = input_image.size
  mask = np.zeros((height, width), dtype=np.uint8)
  h_start, h_end = height // 4, 3 * height // 4
  w_start, w_end = width // 4, 3 * width // 4
  mask[h_start:h_end, w_start:w_end] = 255
  mask_image = Image.fromarray(mask)

  # Generate an inpainted image
  prompt = "a woman with pale skin in a black shirt, oil painting style"
  strength = 0.75

  print(f"\nGenerating inpainted image with prompt: {prompt}")
  print(f"Strength: {strength}")

  image = pipe(
      prompt=prompt,
      image=input_image,
      mask_image=mask_image,
      strength=strength,
      num_inference_steps=8,
      guidance_scale=1.0,
      generator=torch.Generator(device="cuda").manual_seed(42),
  ).images[0]

  output_path = "test_zimage_inpaint_output.png"
  image.save(output_path)
  print(f"\nImage saved to {output_path}")

```

</details>

LoRA functionality is also supported (inherited from ZImageLoraLoaderMixin).

Clipboard3

## Before submitting
- [X] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [X] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [X] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu @asomoza @sayakpaul

### CODE REVIEW

File: src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py [hunk 1]
```diff
@@ -0,0 +1,846 @@
...
+            masked_image_latents = retrieve_latents(self.vae.encode(masked_image), generator=generator)
+
+        # Apply VAE scaling
+        masked_image_latents = (masked_image_latents - self.vae.config.shift_factor) * self.vae.config.scaling_factor
+
+        # Expand for batch size
+        if mask.shape[0] < batch_size:
+            mask = mask.repeat(batch_size // mask.shape[0], 1, 1, 1)
+        if masked_image_latents.shape[0] < batch_size:
+            masked_image_latents = masked_image_latents.repeat(batch_size // masked_image_latents.shape[0], 1, 1, 1)
```
[the remark this answers is not in the document]
Comment (CalamitousFelicitousness) [reply src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py:377]: Added batch size validation using diffusers convention from existing code rather than `!= 0:`

File: src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py [hunk 2]
```diff
@@ -0,0 +1,846 @@
...
+
+                if callback_on_step_end is not None:
+                    callback_kwargs = {}
+                    for k in callback_on_step_end_tensor_inputs:
+                        callback_kwargs[k] = locals()[k]
+                    callback_outputs = callback_on_step_end(self, i, t, callback_kwargs)
+
+                    latents = callback_outputs.pop("latents", latents)
+                    prompt_embeds = callback_outputs.pop("prompt_embeds", prompt_embeds)
+                    negative_prompt_embeds = callback_outputs.pop("negative_prompt_embeds", negative_prompt_embeds)
```
[the remark this answers is not in the document]
Comment (CalamitousFelicitousness) [reply src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py:905]: Done

**Comment by CalamitousFelicitousness** [comment]
@yiyixuxu Ready for re-review.

File: src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py [hunk 3]
```diff
@@ -0,0 +1,860 @@
...
+from ...loaders import FromSingleFileMixin, ZImageLoraLoaderMixin
+from ...models.autoencoders import AutoencoderKL
+from ...models.transformers import ZImageTransformer2DModel
+from ...pipelines.pipeline_utils import DiffusionPipeline
+from ...schedulers import FlowMatchEulerDiscreteScheduler
+from ...utils import logging, replace_example_docstring
+from ...utils.torch_utils import randn_tensor
+from .pipeline_output import ZImagePipelineOutput
+
+
```
Comment (asomoza) [inline src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py:31]:
```suggestion

if is_torch_xla_available():
    import torch_xla.core.xla_model as xm

    XLA_AVAILABLE = True
else:
    XLA_AVAILABLE = False
    
```

you skipped the TPU part here

File: src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py [hunk 4]
```diff
@@ -0,0 +1,860 @@
...
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import ZImageInpaintPipeline
+        >>> from diffusers.utils import load_image
+
+        >>> pipe = ZImageInpaintPipeline.from_pretrained("Z-a-o/Z-Image-Turbo", torch_dtype=torch.bfloat16)
```
Comment (asomoza) [inline src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py]:
```suggestion
        >>> pipe = ZImageInpaintPipeline.from_pretrained("Tongyi-MAI/Z-Image-Turbo", torch_dtype=torch.bfloat16)
```

lets use the official repo here, also the example produces a really bad image, maybe we could use something more updated?

Comment (CalamitousFelicitousness) [reply src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py]: Oh, I'm not sure how that snuck in, I had issues pulling from the original and temporarily swapped it, must've forgotten to clean up.

Regarding the example it looks fine to me, but if that's a blocker I can change it.

Comment (asomoza) [reply src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py]: no, it's not a blocker, just a suggestion, if it looks fine to you, it's fine

File: src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py (hunk 3 above)
Comment (CalamitousFelicitousness) [reply src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py:31]: Good catch, my bad

**Comment by CalamitousFelicitousness** [comment]
@asomoza How does this look? I still have to do cleanup, but in general structure.

**Comment by asomoza** [comment]
looks good to me, let me know when you do the cleanup, also I wonder if we should do the fix for this [issue](https://github.com/huggingface/diffusers/issues/12905) here too. 

ccing: @yiyixuxu

**Comment by CalamitousFelicitousness** [comment]
Might as well, I'll integrate it, can always revert and separate into another PR if preferred.

**Comment by CalamitousFelicitousness** [comment]
@asomoza Apologies for the delay, cleanup and changes requested done.

File: src/diffusers/pipelines/z_image/pipeline_z_image_controlnet.py [hunk 5]
```diff
@@ -365,7 +365,7 @@ def guidance_scale(self):
 
     @property
     def do_classifier_free_guidance(self):
-        return self._guidance_scale > 1
+        return self._guidance_scale > 0
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/z_image/pipeline_z_image_controlnet.py]: @JerryWu-code 
is this change ok? see more context here https://github.com/huggingface/diffusers/issues/12905

Comment (CalamitousFelicitousness) [reply src/diffusers/pipelines/z_image/pipeline_z_image_controlnet.py]: @yiyixuxu I can remove it and we can go ahead with merging of this PR, no point in waiting around, this can be patched at any point.

**Comment by CalamitousFelicitousness** [comment]
@asomoza @yiyixuxu I have removed that commit, ready for final(?) review.

**Comment by vladmandic** [comment]
@yiyixuxu @sayakpaul @asomoza any chance we can prioritize review and merge inpaint pipelines? in general, those are highly demanded and very commonly used, but still often overlooked when adding support for new models.

File: src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py [hunk 6]
```diff
@@ -0,0 +1,932 @@
...
+            A text encoder model to encode text prompts.
+        tokenizer ([`AutoTokenizer`]):
+            A tokenizer to tokenize text prompts.
+        transformer ([`ZImageTransformer2DModel`]):
+            A ZImage transformer model to denoise the encoded image latents.
+    """
+
+    model_cpu_offload_seq = "text_encoder->transformer->vae"
+    _optional_components = []
+    _callback_tensor_inputs = ["latents", "prompt_embeds", "mask", "masked_image_latents"]
```
Comment (asomoza) [inline src/diffusers/pipelines/z_image/pipeline_z_image_inpaint.py:184]: you're adding the `mask` and the `masked_image_latents`, but does this have an applicable use case? in which use case we can use them on each step? Not a blocker but just curious

**Review by asomoza — APPROVED** [review]

**Comment by asomoza** [comment]
thanks, if we do the change to the cfg, we can do it to all the pipelines at the same time later.
